### PR TITLE
Added visdom logger handler.

### LINKED ISSUE #301: Discussion about Metrics plotting
Hello, I would like to discuss the automation the metric plotting.

## Motivation

Currently a big part of (my?) ignite scripts go into the boiler plate of logging the training and testing results, be it a set of **visdom** windows, **mlflow** experiments, logging files or any combination of these. Each time a metric is added or removed, all the logging mechanism needs to be updated accordingly.

## Solution

A possible solution is to have a mechanism similar to the `ProgressBar` (#296), where you create a `Plotting` object (e.g. a wrapper to a **visdom** window or **mlflow** server) and `attach` metrics to it. This object will take care of plotting the metric in the correct context.

An example use case is to have a **visdom** window that plots loss values every iteration or metric values every epoch. Another use case is to log the metrics to an **mlflow** experiment. 

The benefit of this idea is that it will reduce much of the plotting boiler plate that currently has to be "manually" added to event handlers.

## Implementation Discussion

The problem that I see with the `attach` approach (used in the `ProgressBar`) is that it is not clear whether the `EPOCH_COMPLETED` will trigger first the `self.completed` method of the metric, or the plotting callback of the `Plotting` object. In the second case, the metric value will not be ready for plotting.

In my opinion a better method is to add a callback mechanism to the `Metric` class that enables calling arbitrary callbacks (**visdom** plot etc) at the end of the `completed` method.

Another option is to add a wrapper metric (like `RunningAverage`) that will add the callback functionality.

Will appreciate any ideas.

### PR DESCRIPTION
**Pull request by amitibo**
Handler that plots to [Visdom](https://github.com/facebookresearch/visdom).

graphs

Fixes #301
Replaces #304 

Description:
This handler introduces API to create attach visdom plots to metrics and parameters of the training.

Modified the mnist_with_visdom.py to use the new handler.
Added installation of typing module to .travis.yml

Check list:
* [ ] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by amitibo** [comment]
Any ideas how I can make `codecov` happy?

**Comment by vfdev-5** [comment]
@amitibo take a look [here](https://codecov.io/gh/pytorch/ignite/compare/2f48a098cae1208ecf1a5c0e78f3583e6b3b4d2d...f9e866b66d05eda2cb3a79376af9ff9fe37652ef/tree/ignite/contrib/handlers), it says that `visdom_logger.py ` is covered by 15%.
and similar message is shown in github:
```
codecov/patch — 16% of diff hit (target 90%)
```
Maybe we need to improve the data sent to codecov to see which part of code is not covered. The same stuff you can do locally by running something like :
```
py.test --cov ignite --cov-report term-missing tests/ignite/contrib/handlers/visdom_logger.py
```

**Comment by vfdev-5** [comment]
@amitibo maybe we could use `mock` to emulate visdom server behaviour instead of launching real server ?

**Comment by vfdev-5** [comment]
@amitibo I would like to discuss about how to make unified API between this logger and tensorboard logger. 

In some sense, at first glance window abstraction seems to be unnecessary if we do not want to draw in the same window, e.g. train loss, learning rate, ...
On the other hand, drawing metrics of train and val in the same window looks great and this is achieved with this window abstraction. In tensorboard it seems to me we can not draw two curves (train, val) like this in the same window. But in tensorboard we can compare curves of different experiments. Is it possible with Visdom ?


cc @anmolsjoshi

**Comment by amitibo** [comment]
> @amitibo maybe we could use `mock` to emulate visdom server behaviour instead of launching real server ?

I agree but it will have to wait till I have more time.

**Comment by amitibo** [comment]
> @amitibo I would like to discuss about how to make unified API between this logger and tensorboard logger.
> 
> In some sense, at the first glance window abstraction seems to be unnecessary if we do not want to draw in the same window, e.g. train loss, learning rate, ...
> On the other hand, drawing metrics of train and val in the same window looks great and this is achieved with this window abstraction.

This is the reason I added the windows abstraction.

> In tensorboard it seems to me we can not draw two curves (train, val) like this in the same window. But in tensorboard we can compare curves of different experiments. Is it possible with Visdom ?

Yes, but it is part of Visdom, not the handler. I think that it also requires that the windows share the same name, i.e. you compare (visdom) environments, not specific graphs.

**Comment by vfdev-5** [comment]
> Did `codecov` see the test file I added (`tests/ignite/contrib/handlers/visdom_logger.py`)?

Seems like it does not see. [ref](https://travis-ci.org/pytorch/ignite/jobs/453737518#L1127)

File: ignite/contrib/handlers/visdom_logger.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+from typing import Callable, Dict, List, Union
+
+
+class _VisdomWindows:
+    """Object encapsulating a Visdom Window.
+
+    Args:
+        vis (Visdom object): Visdom client.
+        env (str): Name of Visdom environment for the graphs.
+        save_by_default (bool): The graphs will be saved by default by the server.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/visdom_logger.py]: This option name is not clear. According to the API https://github.com/facebookresearch/visdom#vissave
this saves some environment of the server. Maybe we could leave it by default as true ?

**Comment by vfdev-5** [comment]
@amitibo I updated travis and the code of the `test_visdom_logger.py` so that the tests pass but not the coverage.

**Comment by amitibo** [comment]
> @amitibo I updated travis and the code of the `test_visdom_logger.py` so that the tests pass but not the coverage.

Does this require more tests to the visdom logger?

File: ignite/contrib/handlers/visdom_logger.py (hunk 1 above)
Comment (amitibo) [reply ignite/contrib/handlers/visdom_logger.py]: This option saves the environment (your current plots) on the server side. Without this option if the server is closed (crashes) and opened again, all past work is lost.
I agree that it is better to have it True by default.

**Comment by vfdev-5** [comment]
Yes, more tests needed

**Comment by amitibo** [comment]
> Yes, more tests needed

Maybe its better to first decide on the final API (considering also #265), before tests. What do you think @vfdev-5 ?

**Comment by anmolsjoshi** [comment]
Closing this in favor of #455
